[FFL-2679] Add APM trace enrichment guide for Feature Flags (#38175)
DataDog/documentation: content/en/feature_flags/guide/apm_trace_enrichment.md

4. The Datadog backend decodes these tags and writes human-readable `@feature_flags.<flag_key>` facets to the span, making them searchable in Trace Explorer.

The SDK-side tags are transport-only and are stripped server-side. The tags visible to you in Trace Explorer are the decoded `@feature_flags.<flag_key>` facets.

## Enable APM trace enrichment

Set the following environment variables. The first enables the Datadog Feature Flags provider (if not already set), and the second enables span enrichment:

{{< code-block lang="bash" >}}
# Required: Enable the Datadog Feature Flags provider (if not already set)
DD_EXPERIMENTAL_FLAGGING_PROVIDER_ENABLED=true

# Enable APM trace enrichment
DD_EXPERIMENTAL_FLAGGING_PROVIDER_SPAN_ENRICHMENT_ENABLED=true
{{< /code-block >}}

The enrichment environment variable is supported by all supported server-side SDKs. No code changes are required. Enabling the variable activates the enrichment hook automatically when the Datadog OpenFeature provider initializes. Node.js additionally supports code-level configuration as shown in the language tabs below.

### Language-specific configuration

{{< tabs >}}
{{% tab "Node.js" %}}

> 

Reviewer: Sorry, one more suggested tweak: could you reorder the tabs alphabetically by language name? That will match the order in the minimum version table above, and it's a natural order for readers to expect.

Author: Done. Tabs now ordered alphabetically: Go, Java, Node.js — matches the version table.

---
title: Set Up APM Trace Enrichment for Feature Flags
description: Automatically attach feature flag evaluation data to APM traces so you can filter traces by flag variant and use APM metrics in experiments.
further_reading:
- link: "/feature_flags/server/"
  tag: "Documentation"
  text: "Server-Side Feature Flags"
- link: "/feature_flags/guide/server_flag_evaluation_metrics/"
  tag: "Guide"
  text: "Set Up Server-Side Flag Evaluation Metrics"
- link: "/experiments/"
  tag: "Documentation"
  text: "Experiments"
- link: "/tracing/trace_explorer/"
  tag: "Documentation"
  text: "Trace Explorer"
---

## Overview

> APM trace enrichment automatically attaches feature flag evaluation data to your APM traces. When a feature flag is evaluated during a traced request, the SDK records which flags were evaluated and which variants were returned. This data is written to the root span and processed server-side so you can:

Author: > root span
doublechecking this is correct, @sameerank ?

further_reading:
- link: "/feature_flags/server/"
  tag: "Documentation"
  text: "Server-Side Feature Flags"
- link: "/feature_flags/guide/server_flag_evaluation_metrics/"
  tag: "Guide"
  text: "Set Up Server-Side Flag Evaluation Metrics"
- link: "/experiments/"
  tag: "Documentation"
  text: "Experiments"
- link: "/tracing/trace_explorer/"
  tag: "Documentation"
  text: "Trace Explorer"
---

## Overview

APM trace enrichment automatically attaches feature flag evaluation data to your APM traces. When a feature flag is evaluated during a traced request, the SDK records which flags were evaluated and which variants were returned. This data is written to the root span and processed server-side so you can:

- **Filter traces by flag variant** in [Trace Explorer][1] using `@feature_flags.<flag_key>:<variant>` facets.
- **Use APM error rates and latency as experiment metrics** to measure the impact of flag changes on application performance.
- **Debug flag-related issues** by seeing which flags were active when an error occurred.

<div class="alert alert-warning">APM trace enrichment is experimental and may change in a future release.</div>

> <div class="alert alert-info">APM trace enrichment is available in the following SDKs. Support for additional languages is in progress.</div>

Reviewer: It's OK to mention that the feature is experimental, but avoid commenting on WIP features. Also the second alert could be a regular body paragraph:
```suggestion
<div class="alert alert-warning">APM trace enrichment is experimental and may change in a future release.</div>

APM trace enrichment is available in the following SDKs:
```

Author: Removed the alert-info div. SDK list is now a plain paragraph and table.

| Language | Status | Minimum version |
| -------- | ------ | --------------- |
| Go       | Available | 2.8.0        |

> | Java     | Available | _TODO_       |

Reviewer: Actually, we should avoid "Coming soon" information altogether. We should document the current state of the product, not what we anticipate for the future. Remove any as-yet-unsupported languages for now, and we'll add them back when they're supported.

That means we can also remove the "Status" column, because we'll only document functionality that's "Available."

> | Java     | Available | _TODO_       |

Author: Fixed — only showing what's shipped! Good pragma.

| Node.js  | Available | 5.105.0      |

> | Python   | Available | _TODO_       |

Reviewer: Same comment as above: remove any as-yet-unsupported languages for now

> | Python   | Available | _TODO_       |

Author: Addressed in the same commit. Python removed from the table until 4.12.0 releases.

| .NET     | Coming soon | —          |
| PHP      | Coming soon | —          |

> | Ruby     | Coming soon | —          |

Reviewer: Go with option # 2 here: remove the code example tabs for any languages that aren't yet supported. We can add them back when the languages are available.

> | Ruby     | Coming soon | —          |

Author: Went with option 2. Removed .NET, PHP, Ruby, and Python tabs. Only Go, Java, and Node.js tabs remain.

## Prerequisites

Before setting up APM trace enrichment, confirm the following:

- [Server-side feature flags][2] are already configured and flags are evaluating in your application.
- [APM tracing][3] is enabled and traces are flowing to Datadog.

> - `DD_EXPERIMENTAL_FLAGGING_PROVIDER_ENABLED=true` is set, or the equivalent code-based configuration for your language. See [Server-Side Feature Flags][2] for setup details.

Reviewer: The two identical links to server-side flags seem repetitive in a short section like this. How about:
```suggestion
- Server-side feature flags are already configured and flags are evaluating in your application.
- [APM tracing][3] is enabled and traces are flowing to Datadog.
- `DD_EXPERIMENTAL_FLAGGING_PROVIDER_ENABLED=true` is set, or the equivalent code-based configuration for your language. 

See [Server-Side Feature Flags][2] for details on setting up flags for supported languages. 
```

Author: Applied the suggestion. Single link at the end of the prerequisites section.

> ## How it works

Reviewer: Detailed headings are better for AI (and human) consumption:
```suggestion
## How APM trace enrichment works
```

Author: Applied. Now reads "How APM trace enrichment works."

1. Each time a flag is evaluated, the SDK captures the evaluation metadata (flag serial ID, targeting key, and default fallback value).
2. The metadata is accumulated on the root span of the current trace.
3. When the root span finishes, the SDK writes the accumulated data as compact span tags (`ffe_flags_enc`, `ffe_subjects_enc`, `ffe_runtime_defaults`).
4. The Datadog backend decodes these tags and writes human-readable `@feature_flags.<flag_key>` facets to the span, making them searchable in Trace Explorer.

The SDK-side tags are transport-only and are stripped server-side. The tags visible to you in Trace Explorer are the decoded `@feature_flags.*` facets.

## Enable APM trace enrichment

Set the following environment variables. The first enables the Datadog Feature Flags provider (if not already set), and the second enables span enrichment:

{{< code-block lang="bash" >}}
# Required: Enable the Datadog Feature Flags provider (if not already set)
DD_EXPERIMENTAL_FLAGGING_PROVIDER_ENABLED=true

# Enable APM trace enrichment
DD_EXPERIMENTAL_FLAGGING_PROVIDER_SPAN_ENRICHMENT_ENABLED=true
{{< /code-block >}}

> The enrichment environment variable is supported by all server-side SDKs. No code changes are required. Enabling the variable activates the enrichment hook automatically when the Datadog OpenFeature provider initializes. Node.js and Ruby additionally support code-level configuration as shown in the language tabs below.

Author: Removed Ruby reference. Changed "all server-side SDKs" to "all supported server-side SDKs."

No additional code configuration is needed. The `DD_EXPERIMENTAL_FLAGGING_PROVIDER_SPAN_ENRICHMENT_ENABLED` environment variable enables span enrichment. The PHP SDK supports enrichment through both the OpenFeature `DataDogProvider` and the native `DDTrace\FeatureFlags\Client`.

{{< code-block lang="php" filename="app.php" >}}
<?php

require_once __DIR__ . '/vendor/autoload.php';

use DDTrace\OpenFeature\DataDogProvider;
use OpenFeature\OpenFeatureAPI;

$api = OpenFeatureAPI::getInstance();
$api->setProvider(new DataDogProvider());

$client = $api->getClient('my-service');
// Flag evaluations now enrich APM spans automatically
{{< /code-block >}}

{{% /tab %}}
{{% tab ".NET" %}}

No additional code configuration is needed. The `DD_EXPERIMENTAL_FLAGGING_PROVIDER_SPAN_ENRICHMENT_ENABLED` environment variable enables span enrichment.

{{< code-block lang="csharp" filename="Program.cs" >}}
using OpenFeature;
using Datadog.FeatureFlags.OpenFeature;

var provider = new DatadogProvider();
await Api.Instance.SetProviderAsync(provider);

var client = Api.Instance.GetClient("my-service");
// Flag evaluations now enrich APM spans automatically
{{< /code-block >}}

{{% /tab %}}
{{< /tabs >}}

> ## Verify enrichment is working

Reviewer: ```suggestion
## Verify APM trace enrichment
```

Author: Applied. Now reads "Verify APM trace enrichment."

After deploying with span enrichment enabled:

1. Trigger requests in your application that evaluate feature flags.
2. Go to [Trace Explorer][1] and search for a recent trace from your service.
3. Open a trace and look for `@feature_flags.*` attributes on the root span.

> The SDK writes compact encoded tags (`ffe_flags_enc`, `ffe_subjects_enc`, `ffe_runtime_defaults`) to the root span. The Datadog backend decodes these and produces human-readable `@feature_flags.*` facets. This processing takes a few seconds after the span is ingested.

Reviewer: This follows the same pattern used above in "How it works." Ignore this if we actually want users to use the `*` wildcard in their searches (though I'm guessing that searching for their specific flag key would give a more targeted results than a wildcard).
```suggestion
3. Open a trace and look for `@feature_flags.<flag_key>` attributes on the root span.

The SDK writes compact encoded tags (`ffe_flags_enc`, `ffe_subjects_enc`, `ffe_runtime_defaults`) to the root span. The Datadog backend decodes these and produces human-readable `@feature_flags.<flag_key>` facets. This processing takes a few seconds after the span is ingested.
```

Author: Applied. Step 3 now reads `@feature_flags.<flag_key>`.

After backend processing, the root span contains attributes like the following:

| Attribute | Example value |
| --------- | ------------- |
| `@feature_flags.checkout-flow` | `treatment` |

> | `@feature_flags.dark-mode` | `control` |

Reviewer: Really make it clear these are just examples:
```suggestion
After backend processing, the root span contains attributes such as the following examples:

| Example attribute | Example value |
| --------- | ------------- |
| `@feature_flags.checkout-flow` | `treatment` |
| `@feature_flags.dark-mode` | `control` |
```

Author: Applied. Columns now read "Example attribute" and "Example value."

Each attribute key is `@feature_flags.<flag_key>` and the value is the variant returned by the evaluation.

### Troubleshooting

If `@feature_flags.*` attributes do not appear on your traces:

- Confirm the Datadog OpenFeature provider is active (`DD_EXPERIMENTAL_FLAGGING_PROVIDER_ENABLED=true`).
- Confirm span enrichment is enabled (`DD_EXPERIMENTAL_FLAGGING_PROVIDER_SPAN_ENRICHMENT_ENABLED=true`).
- Verify that your application is evaluating flags during traced requests. Enrichment only occurs when a flag is evaluated while a trace is active.
- Wait a few seconds after the span is ingested. The `@feature_flags.*` facets are derived by backend processing and do not appear in raw span metadata.

> - For debugging, inspect raw span metadata for the `ffe_flags_enc` tag. If this tag is present, the SDK is emitting enrichment data and the backend has not processed it yet or the feature flag gate is not enabled for your organization.

Reviewer: ```suggestion
- For debugging, inspect raw span metadata for the `ffe_flags_enc` tag. If this tag is present, the SDK is emitting enrichment data. Either the backend has not processed it yet, or the feature flag gate is not enabled for your organization.
```

Author: Applied the suggestion. Split into two sentences.

> ### Search and filter by flag variant

Reviewer: This section and the next one seem to be standalone topics, not sub-parts of the "Verify it's working" heading above. If that's the case, promote to h2:
```suggestion
## Search and filter by flag variant
```

Author: Promoted to H2.

Use the `@feature_flags.*` facets to filter traces in Trace Explorer:

> | Use case | Query |

Reviewer: ```suggestion
These examples use the `@feature_flags.<flag_key>` facets to filter traces in Trace Explorer:

| Use case | Example query |
```
```suggestion
Use the `@feature_flags.*` facets to filter traces in Trace Explorer:

| Use case | Query |
```

Author: Applied. "These examples use..." and column renamed to "Example query."

| -------- | ----- |
| Traces for a specific variant | `@feature_flags.checkout-flow:treatment` |
| Errors under a variant | `@feature_flags.checkout-flow:treatment status:error` |
| Any trace where a flag was evaluated | `@feature_flags.checkout-flow:*` |
| Multiple flags on the same request | `@feature_flags.checkout-flow:treatment @feature_flags.new-search:enabled` |
| Scoped to service and environment | `env:production service:api-gateway @feature_flags.rate-limit-v2:enabled` |

> ### Use enriched traces across Datadog

Reviewer: ```suggestion
## Use enriched traces across Datadog
```

Author: Promoted to H2.

Feature flag attributes on traces are available across Datadog:

- **Monitors**: Alert when the error count for a specific variant exceeds a threshold to catch variant-specific regressions.
- **Dashboards**: Add a timeseries widget comparing p99 latency across variants using `@feature_flags.<flag_key>` as a group-by dimension.
- **Notebooks**: Build an investigation notebook comparing control and treatment performance during an experiment.
- **Visualizations**: Use the Top List view in Trace Explorer to verify that rollout traffic distribution matches your targeting rules.

## Use with experiments

> APM trace enrichment enables APM-derived metrics as experiment metrics. After enrichment is enabled, the Datadog Experiments platform correlates feature flag allocations with trace-level data. You can measure the impact of flag variants on:

Reviewer: ```suggestion
APM trace enrichment enables APM-derived metrics as experiment metrics. After enrichment is enabled, Datadog Experiments correlate feature flag allocations with trace-level data. You can measure the impact of flag variants on:
```

Author: Applied. Dropped "the" before "Datadog Experiments."
